### DISCUSSION TITLE
ImageMagick Temp File

### DISCUSSION THREAD

**Discussion by programmerarunkumar** [ImageMagick/ImageMagick#7167] [Help]
### ImageMagick version

7.1.0-6

### Operating system, version and so on

CentOS,7

### Description

Im using below MagickWand methods for imageresize
NewMagickWand()
MagickReadImage()
MagickResizeImage()
MagickWriteImage()
DestroyMagickWand()

some times magick-* file in the temp-folder  is not cleared.

can some help me to find the root cause ?

**Comment by snibgo** [comment]
If your program exits in an unexpected manner (it crashes, or is killed by the OS) then it has no opportunity to remove temporary files.

**Reply by programmerarunkumar** [reply]
what could be possible reason for unexpected crash or kill ?

**Comment by snibgo** [comment]
There are many possible reasons, including bugs in your program.

**Reply by programmerarunkumar** [reply]
In my case mutliple thread will access the above mentioned magickwand api SINCE MAGICKWAND IS THREADSAFE.
Multiple Thread accessing the the files in the TEMP_PATH can cause any issue ?

**Comment by urban-warrior** [comment]
ImageMagick deletes any temporary file resources it creates before the program exits. If any temporary files remain, it could be a result of a signal that ImageMagick cannot intercept such as SIGBUS. To prevent a SIGBUS signal from occurring, use this security policy:

```
<policy domain="cache" name="synchronize" value="True"/>
```

Set to `True` to ensure all image data is fully flushed and synchronized to disk. There is a performance penalty, however, the benefits include ensuring a valid image file in the event of a system crash and early reporting if there is not enough disk space for the image pixel cache.

The only other reason for left-over temporary files might be a bug.  To investigate, we would need a method to reproduce the bug.

**Comment by programmerarunkumar** [comment]
>To investigate, we would need a method to reproduce the bug.

NewMagickWand()
MagickReadImage()
MagickResizeImage()
MagickWriteImage()
DestroyMagickWand()
Im using the above method to resize the image 

> however, the benefits include ensuring a valid image file in the event of a system crash

what do you mean by ensuring a valid image file  here ?

> There is a performance penalty

could you explain more about this ?

**Reply by urban-warrior** [reply]
To investigate further, we would need you post a MagickWand program we can download, compile, and run that illustrates the issue.  In the meantime, do you call MagickWandTerminus()?  It is a function that is used to clean up and release resources when shutting down an application that uses ImageMagick. This function should be called in the primary thread of the application's process during the shutdown process. It's crucial that this function is invoked only after any threads that are using ImageMagick functions have terminated.

**Reply by programmerarunkumar** [reply]
sure i will share my code 
Also Im not using MagickWandTerminus() since imagemagick will be used all time in my case.

**Reply by programmerarunkumar** [reply]
[My MagickWand Program](https://drive.google.com/file/d/12bimg38hOZQZOg38Umg-Ee7SyxTomBVV/view?usp=sharing)

**Comment by urban-warrior** [comment]
Thank you for reporting the issue. We have successfully reproduced it and are actively working on a patch to resolve it. You can expect this patch to be merged into the [main GIT branch](https://github.com/ImageMagick/ImageMagick), later today. As part of our commitment to quality, this fix will also be included in the upcoming [beta releases of ImageMagick](https://imagemagick.org/archive/beta/) by tomorrow. Your patience and feedback are greatly appreciated.

**Reply by programmerarunkumar** [reply]
In the latest ImageMagick. I'm not having the magick* in temp folder even if the disk resource is exhausted. Thank you
